Kyle/symco 18

kaigoos opened the pull request:
Added a test to make sure when a template is deleted all related template node, template node choice, template response, and template node response objects are deleted.
Also changed the deletion modal to display the template name correctly.

kaigoos committed:
Making sure template deletion also deletes related tempalte node,
template node choice, template response, and template node response
objects to be deleted. Also minor changes to deletion modal.

kaigoos committed:
Merge branch 'main' into kyle/SYMCO-18

rosasck commented on simcon_project/conversation_templates/tests/test_template_management.py:
```diff
@@ -1,10 +1,11 @@
+from ..models import *
 from ..forms import FolderCreationForm
-from ..views.template_management import route_to_current_folder, remove_template, FolderEditView
+from ..views.template_management import route_to_current_folder, FolderEditView
```
Route to current folder, should not be snake case. Not sure if this was you or before

kaigoos replied:
remove_template shouldn't be snake case either then right?

rosasck commented on simcon_project/templates/template_management/template_delete_modal.html:
```diff
@@ -10,8 +10,8 @@ <h3 class="modal-title">Delete Template</h3>
-    <p>Are you sure you want to delete
-      <strong>{{ template.name }}</strong>?</p>
+    <p>Are you sure you want to delete <strong>{{ template_name }}</strong>?</p>
+    <p>All data related to this template will be deleted permanently.</p>
```
"All responses and assignments that include this template will be deleted permanently"

kaigoos replied:
Wait are the assignments suppose to be deleted too?

rosasck reviewed (dismissed):
minor fixes

JerCarnahan reviewed (dismissed):
All good aside from things already pointed out

kaigoos committed:
Added test to make sure templates are removed from the assignments they
are in

kaigoos committed:
Made it so deleting a template will also remove assignments that have
the template as it's only template
Didn't write a test for this since it would be out of scope

kaigoos committed:
Made it so deleting a template will also remove assignments that have
the template as it's only template
Didn't write a test for this since it would be out of scope

kaigoos committed:
Merge branch 'kyle/SYMCO-18' of https://github.com/Likhovodov/Simulated-Conversations into kyle/SYMCO-18

kaigoos commented:
Added a section in the test to make sure the template being deleted was also removed from all assignments that contain it.
Also made it so if an assignment only has one template and that template is being deleted then the assignment is also deleted.

devin-w-hill commented on simcon_project/conversation_templates/views/template_management.py:
```diff
@@ -141,8 +141,64 @@ class TemplateDeleteView(BSModalDeleteView):
 
-
-def remove_template(request, pk):
+    def get(self, request, *args, **kwargs):
```
Maybe I am missing something but I think you have "get" defined twice here? Lines 144 and 172

devin-w-hill replied:
Post as well. I can't see any differences but I'm not sure if there's some magic going on here.

Likhovodov commented on simcon_project/conversation_templates/tests/test_template_management.py:
```diff
@@ -1,10 +1,11 @@
+from ..views.template_management import RouteToCurrentFolder, FolderEditView
 
 
 class TemplateManagementTests(TestCase):
```
What about the cleanup of the objects created in this suit? test_delete_template_cascades probably takes care of some or all of that, but I'm not sure. Also, I don't know anything about tests in Django haha

kaigoos replied:
I'll look into it and see how TestCase takes care of the objects that were created within the class

kaigoos replied:
So django creates a database specifically for testing then destroys it once the test is done so we don't have to worry about the objects that were created during the test

Likhovodov commented on simcon_project/conversation_templates/views/template_management.py:
```diff
@@ -141,8 +141,64 @@ class TemplateDeleteView(BSModalDeleteView):
+        context = {"template_name": this_template.name, "assignments": to_delete}
+        return render(request, self.template_name, context)
+
+    def post(self, request, *args, **kwargs):
```
Looks like you might have forgotten to delete a post after moving things around (post is already defined above)

kaigoos replied:
Something got really messed up here since get and post are in there twice and I don't even think it's in the right spot.

Likhovodov reviewed (commented):
Looking good so far

kaigoos committed:
Minor PR changes

devin-w-hill reviewed (dismissed):
Lookin good!

kaigoos committed:
Merge branch 'main' into kyle/SYMCO-18

rosasck commented on simcon_project/templates/researcher_base.html:
```diff
@@ -18,6 +18,7 @@
         </div>
         <ul class="nav navbar-nav">
           <li>{% block nav-researcherView %}<a href="{% url 'ResearcherView' %}">Home</a>{% endblock %}</li>
+          <li>{% block nav-templateView %}<a href="{% url 'management:main' %}">Templates</a>{% endblock %}</li>
```
You should be able to just say "MainView".. but this seems to work!

kaigoos replied:
I used the alias since I thought management:main would be more clear than just MainView

kaigoos committed:
Template management is unique to researcher now

kaigoos commented:
Updated so template management can be accessed through researcher view and the contents of the page (TemplateFolder and ConversationTemplate objects) that are displayed are unique to the researcher. 
- Folders and templates are unique to each researcher
- Redirects to login if the user is not a researcher and tries to access template management page
- Deleting a template deletes all relevant data (TemplateNode, TemplateNodeResponse, TemplateResponse, TemplateNodeChoice, Assignments-if there is only one template and it is being deleted)
- Confirmation modal appears with relevant message when deleting a template
- Confirmation modal lists assignments that will also be deleted when deleting the only template in assignment
- Also made changes to TemplateFolder model

ericharper1 commented on simcon_project/conversation_templates/forms.py:
```diff
@@ -20,6 +21,10 @@ class FolderCreationForm(BSModalModelForm):
+        # Only show templates unique to each user in the form.
+        if self.request.user:
+            templates = ConversationTemplate.objects.filter(researcher=self.request.user.id)
+            self.fields['templates'].queryset = templates.all()
```
Is there a specific reason to call `.all()`?

kaigoos replied:
That's true, it probably works without it since I'm already using filter in the line above it

ericharper1 commented on simcon_project/conversation_templates/views/template_management.py:
```diff
@@ -141,8 +154,36 @@ class TemplateDeleteView(BSModalDeleteView):
+        return redirect(reverse('management:main'))
+
+
+def RemoveTemplate(request, pk):
```
Functions should be snake_cased. At some point, I think we'll do a sweep of all the view functions to properly case them.

kaigoos replied:
I'm getting conflicting info, I used to have it snake case since pycharm tells me to, but I was told to change it to this way in another PR

ericharper1 replied:
Hmm. I'm going with the Django/Python docs on this one

ericharper1 replied:
I'll fight whoever says otherwise lol

devin-w-hill replied:
I wish you had told me when I had commented it on all your files Eric! haha

ericharper1 commented on simcon_project/conversation_templates/views/template_management.py:
```diff
@@ -158,7 +199,7 @@ def remove_template(request, pk):
 
 
-def route_to_current_folder(previous_url):
+def RouteToCurrentFolder(previous_url):
```
snake_case

ericharper1 commented on simcon_project/conversation_templates/views/template_management.py:
```diff
@@ -71,7 +78,8 @@ def main_view(request):
 
-def folder_view(request, pk):
+@user_passes_test(is_researcher)
+def FolderView(request, pk):
```
snake_case

ericharper1 commented on simcon_project/conversation_templates/views/template_management.py:
```diff
@@ -48,15 +50,20 @@ class Meta:
+
+
+@user_passes_test(is_researcher)
+def MainView(request):
```
snake_case

ericharper1 reviewed (changes requested):
It wouldn't let me comment on the Pipfile, but you can delete that. Looks great other than a couple nit-picks!

kaigoos committed:
PR changes

ericharper1 reviewed (approved):
Noice

devin-w-hill reviewed (approved):
Sorry it took so long to test everything out. I didn't update your branch on my end so I wasted a bit of time and had to restart. I went down the list you made in your comment and everything works as you stated, didn't notice any bugs. The css for the buttons on the template management view could use some work but functionally everything is great. Approving but agree with Eric that functions should be snake case if that's what the standard is.
